classes/patreon_login.php

```diff
@@ -14,130 +14,234 @@ public static function updateExistingUser($user_id, $user_response, $tokens) {
 		update_user_meta($user_id, 'patreon_refresh_token', $tokens['refresh_token']);
 		update_user_meta($user_id, 'patreon_access_token', $tokens['access_token']);
 		update_user_meta($user_id, 'patreon_user', $user_response['data']['attributes']['vanity']);
+		update_user_meta($user_id, 'patreon_user_id', $user_response['data']['id']);
+		update_user_meta($user_id, 'patreon_last_logged_in', time());
 		update_user_meta($user_id, 'patreon_created', $user_response['data']['attributes']['created']);
 		update_user_meta($user_id, 'user_firstname', $user_response['data']['attributes']['first_name']);
 		update_user_meta($user_id, 'user_lastname', $user_response['data']['attributes']['last_name']);
 		update_user_meta($user_id, 'patreon_token_minted', microtime());
+		update_user_meta($user_id, 'patreon_token_expires_in', $tokens['expires_in']);
 
 	}
 
-	public static function updateLoggedInUser($user_response, $tokens, $redirect = false) {
+	public static function updateLoggedInUserForStrictoAuth($user_response, $tokens, $redirect = false) {
 
 		$user = wp_get_current_user();
 
 		if(0 == $user->ID) {
+			$redirect = add_query_arg( 'patreon_message', 'patreon_cant_login_strict_oauth', $redirect);
+			wp_redirect($redirect);
+			exit;			
 		} else {
 			/* update user meta data with patreon data */
 			self::updateExistingUser($user->ID, $user_response, $tokens);
+			wp_redirect($redirect);
+			exit;				
 		}
 
 		if($redirect == false || is_null($redirect) ) {
 			wp_redirect(home_url());
 			exit;
 		} else {
-			wp_redirect( get_permalink($redirect) );
+			wp_redirect( $redirect );
 			exit;
 		}
 
 	}
 
-	public static function createUserFromPatreon($user_response, $tokens, $redirect = false) {
+	public static function checkTokenExpiration($user_id=false) {
 
-		global $wpdb;
-
-		$login_with_patreon = get_option('patreon-enable-login-with-patreon', false);
-		$admins_editors_login_with_patreon = get_option('patreon-enable-allow-admins-login-with-patreon', false);
-
-		$email = $user_response['data']['attributes']['email'];
-
-		//use email as login name & do bunch of stuff to it
-		$username = $email;
-		$namesplosion = explode('@', $username, 2);
-		$firstchunk = $namesplosion[0];
-		$username = $firstchunk . base_convert(password_hash($username.time(),PASSWORD_BCRYPT), 8, 32);
-
-		//if login with patreon is enabled
-		if($login_with_patreon == false ) {
-
-			if(username_exists($username)) {
-
-				$suffix = $wpdb->get_var( $wpdb->prepare(
-					"SELECT 1 + SUBSTR(user_login, %d) FROM $wpdb->users WHERE user_login REGEXP %s ORDER BY 1 DESC LIMIT 1",
-					strlen( $username ) + 2, '^' . $username . '(\.[0-9]+)?$' ) );
+		if($user_id) {
+			$user = get_user_by('ID',$user_id);
+		}
+		else {
+			$user = wp_get_current_user();
+		}
 
-				if( !empty( $suffix ) ) {
-					$username .= ".{$suffix}";
+		if(!$user OR 0 == $user->ID) {
```

Comment:
still need to address this empty if. flip the logic perhaps?

Reply:
Yeah, flipping the logic will eliminate the need for placeholder if case. We arent using it for anything currently anyway.

https://github.com/codebard/patreon-wordpress/commit/4df9d72237bf6afff7a12d612ad84baaeb126814